server/grpcsvr/bmc/machine.go

```diff
@@ -19,167 +19,152 @@ type MachineAction struct {
 	StatusMessages    chan string
 }
 
+type power interface {
+	connection() repository.Error
+	close()
+	on() (string, repository.Error)
+	off() (string, repository.Error)
+	status() (string, repository.Error)
+	reset() (string, repository.Error)
+	hardoff() (string, repository.Error)
+	cycle() (string, repository.Error)
+}
+
 // BootDevice functionality for machines
-func (p MachineAction) BootDevice() (string, repository.Error) {
-	var result string
-	errMsg := repository.Error{
-		Code:    0,
-		Message: "",
-		Details: nil,
-	}
-	l := p.Log.GetContextLogger(p.Ctx)
-	l.V(0).Info("not implemented")
-	msg := "power OFF not implemented"
-	l.V(1).Info(msg)
+func (m MachineAction) BootDevice() (result string, errMsg repository.Error) {
+	l := m.Log.GetContextLogger(m.Ctx)
+	msg := "setting Boot Device not implemented yet"
+	l.V(0).Info(msg)
 	errMsg.Code = v1.Code_value["UNIMPLEMENTED"]
 	errMsg.Message = msg
 	return result, errMsg //nolint
 }
 
+type powerConnection struct {
+	name      string
+	connected bool
+	pwr       power
+	err       repository.Error
+}
+
 // Power functionality for machines
-func (p MachineAction) Power() (string, repository.Error) {
-	l := p.Log.GetContextLogger(p.Ctx)
-	l.V(0).Info("power state")
-	// TODO handle nil values
-	var result string
-	errMsg := repository.Error{
-		Code:    0,
-		Message: "",
-		Details: nil,
+func (m MachineAction) Power() (result string, errMsg repository.Error) {
+	l := m.Log.GetContextLogger(m.Ctx)
+	host, user, password, errMsg := m.parseAuth(m.PowerRequest.Authn)
+	if errMsg.Message != "" {
+		return result, errMsg
 	}
 
-	if p.PowerRequest.Authn == nil || p.PowerRequest.Authn.Authn == nil {
-		msg := "no auth found"
-		errMsg.Code = v1.Code_value["UNAUTHENTICATED"]
-		errMsg.Message = msg
-		return msg, errMsg //nolint
-	}
-	user := p.PowerRequest.GetAuthn().GetDirectAuthn().Username
-	password := p.PowerRequest.GetAuthn().GetDirectAuthn().Password
-	host := p.PowerRequest.GetAuthn().GetDirectAuthn().GetHost().Host
+	base := "power " + m.PowerRequest.GetAction().String()
+	msg := "working on " + base
+	m.sendStatusMessage(msg)
 
-	p.StatusMessages <- "trying to connect to bmc"
+	connections := []powerConnection{
+		{name: "ipmi", pwr: &ipmiBMC{mAction: m, user: user, password: password, host: host}},
+		{name: "bmclib", pwr: &bmclibBMC{mAction: m, user: user, password: password, host: host}},
+		{name: "redfish", pwr: &redfishBMC{mAction: m, user: user, password: password, host: host}},
+	}
 
-	connection, err := discover.ScanAndConnect(host, user, password, discover.WithLogger(l))
-	if err != nil {
-		// TODO set errMsg.Code based on err response
+	var connected []bool
```

Comment:
this can just be a bool since we only care if at least one is connected, and setting it would be:
```go
connected = connected || true
```
when we've made a successful connection.

Reply:
sorry, not following. we only care if none are connected because if more than one is connected, we will try to make the power call with all connections, down below in line 98, until one is successful. 

if connected is just a `boolean` and as we're not breaking the loop (line 70) when a connection is successful, I'm not following how we would know if the first out of the 3 connections was successful but not the last 2. in this scenario, connected would be false.

Reply:
Well you still track the connect status of each connection-type in connections (powerConnection.connected). I'm talking about `connected` that is used only in `doesNotContainTrue`. So the new code would be:

```diff
-       var connected []bool
+       var connected bool
        m.sendStatusMessage("connecting to BMC")
        for index := range connections {
                connections[index].err = connections[index].pwr.connection()
                if connections[index].err.Message != "" {
                        connections[index].connected = false
-                       connected = append(connected, false)
                } else {
                        connections[index].connected = true
                        defer connections[index].pwr.close()
-                       connected = append(connected, true)
+                       connected = connected || true
                }
        }
        l.V(1).Info("connections", "connections", fmt.Sprintf("%+v", connections))
-       if doesNotContainTrue(connected) {
+       if !connected {
                m.sendStatusMessage("connecting to BMC failed")
                var combinedErrs []string
                for _, connection := range connections {
```

this `connected` (maybe rename to anyConnected...?) is only used for handling when no connections were successful.

Reply:
also if we do `var connected bool` then we don't need to do the
```go
if  connections....err.Message != "" {
} else { 
}
```
since in the error case .connected is going to be false by default anyway so we only need to handle the happy case with:

```go
if connections[index].err.Message == "" {
    connections[index].connected = true
    defer connections[index].power.close()
    connected = connected || true
}
```

Reply:
ahh, I think I see now. thanks for the explanation. I'll update it.

Reply:
in this case do we even need the `connected = connected || true` couldnt it just be `connected = true`

Reply:
ah right nope don't need that either :D. We'd only want the `|| true` to keep track of *any* false (but really would have been a && anyway).